### PR TITLE
HSET HDEL replace fields w/o reindexing

### PR DESCRIPTION
**Pull request by ashtul**

### CODE REVIEW

File: src/notifications.c [hunk 1]
```diff
@@ -31,16 +44,70 @@ int HashNotificationCallback(RedisModuleCtx *ctx, int type, const char *event, R
...
+
+  // HSET receives field&value, HDEL receives field
+  int cmdFactor = hset ? 2 : 1;
+  int fieldsNum = (numArgs - 2) / cmdFactor;
+  hashFields = (char **)rm_calloc(fieldsNum + 1, sizeof(*hashFields));
+
+  for (size_t i = 0; i < fieldsNum; ++i) {
+    const RedisModuleString *field = RedisModule_CommandFilterArgGet(filter, 2 + i * cmdFactor);
+    const char *fieldStr = RedisModule_StringPtrLen(field, &len);
+    hashFields[i] = rm_strdup(fieldStr);
```
Comment (MeirShpilraien) [inline src/notifications.c:98]: Can we retain the string instead of copy it?

Comment (ashtul) [reply src/notifications.c:98]: @MeirShpilraien 
Tried retaining string. It crashes.
Possibly due to `robj` being on the stack.

File: src/spec.c [hunk 2]
```diff
@@ -1740,14 +1740,36 @@ dict *Indexes_FindMatchingSchemaRules(RedisModuleCtx *ctx, RedisModuleString *ke
   return specs;
 }
 
-void Indexes_UpdateMatchingWithSchemaRules(RedisModuleCtx *ctx, RedisModuleString *key) {
+static bool hashFieldChanged(IndexSpec *spec, char **hashFields) {
```
Comment (rafie) [inline src/spec.c:1743]: Possibly add a TODO comment urging us to optimize the implementation.

**Comment by MeirShpilraien** [comment]
We also need to take care of:
- HINCRBY
- HINCRBYFLOAT
- HSETNX

File: src/notifications.c [hunk 3]
```diff
@@ -59,18 +69,30 @@ int HashNotificationCallback(RedisModuleCtx *ctx, int type, const char *event, R
...
     else CHECK_AND_CACHE_EVENT(rename_to)
+    else CHECK_AND_CACHE_EVENT(trimmed)
+    else CHECK_AND_CACHE_EVENT(restore)
+    else CHECK_AND_CACHE_EVENT(expired)
+    else CHECK_AND_CACHE_EVENT(change)
   }
 
   const char *key_cp = RedisModule_StringPtrLen(key, NULL);
-  if (hset || hmset || hsetnx || hincrby || hincrbyfloat || hdel) {
+  if (hset || hmset || hsetnx || hincrby || hincrbyfloat || hdel || restore) {
```
Comment (MeirShpilraien) [inline src/notifications.c:79]: @ashtul can you add tests for restore and expire?

**Comment by ashtul** [comment]
fix #1395 
fix #1396

File: src/spec.c [hunk 4]
```diff
@@ -1750,14 +1750,38 @@ dict *Indexes_FindMatchingSchemaRules(RedisModuleCtx *ctx, RedisModuleString *ke
...
+    const char *field = RedisModule_StringPtrLen(hashFields[i], NULL);
+    for (size_t j = 0; j < spec->numFields; ++j) {
+      if (!strcmp(field, spec->fields[j].name)) {
+        return true;
+      }
+    }
+    if (!strcmp(field, spec->rule->lang_field) ||
+        !strcmp(field, spec->rule->score_field) ||
+        !strcmp(field, spec->rule->payload_field)) {
+          return true;
```
Comment (MeirShpilraien) [inline src/spec.c:1769]: If the score or the payload changed there is no need to reindex. We can keep it now but just add a comment to optimize it.

**Comment by MeirShpilraien** [comment]
Tests pass on crdt

**Pull request merged**